presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ReorderJoins.java

```diff
@@ -0,0 +1,440 @@
+                if (!(TRUE_LITERAL).equals(filter)) {
+                    planNode = new FilterNode(idAllocator.getNextId(), planNode, filter);
+                }
+                return new JoinEnumerationResult(Optional.of(lookup.getCumulativeCost(planNode, session, symbolAllocator.getTypes())), Optional.of(planNode));
+            }
+            return chooseJoinOrder(nodes, outputSymbols);
+        }
+
+        private static boolean isJoinEqualityCondition(Expression expression)
+        {
+            return expression instanceof ComparisonExpression
+                    && ((ComparisonExpression) expression).getType() == EQUAL
+                    && ((ComparisonExpression) expression).getLeft() instanceof SymbolReference
+                    && ((ComparisonExpression) expression).getRight() instanceof SymbolReference;
+        }
+
+        private static JoinNode.EquiJoinClause toEquiJoinClause(ComparisonExpression equality, Set<Symbol> leftSymbols)
+        {
+            Symbol leftSymbol = Symbol.from(equality.getLeft());
+            Symbol rightSymbol = Symbol.from(equality.getRight());
+            JoinNode.EquiJoinClause equiJoinClause = new JoinNode.EquiJoinClause(leftSymbol, rightSymbol);
+            return leftSymbols.contains(leftSymbol) ? equiJoinClause : equiJoinClause.flip();
+        }
+
+        private JoinEnumerationResult setJoinNodeProperties(JoinNode joinNode)
+        {
+            List<JoinEnumerationResult> possibleJoinNodes = new ArrayList<>();
+            FeaturesConfig.JoinDistributionType joinDistributionType = getJoinDistributionType(session);
+            if (joinDistributionType.canRepartition() && !mustReplicate(joinNode, lookup)) {
+                JoinNode node = joinNode.withDistributionType(PARTITIONED);
+                possibleJoinNodes.add(new JoinEnumerationResult(Optional.of(lookup.getCumulativeCost(node, session, symbolAllocator.getTypes())), Optional.of(node)));
+                node = node.flipChildren();
+                possibleJoinNodes.add(new JoinEnumerationResult(Optional.of(lookup.getCumulativeCost(node, session, symbolAllocator.getTypes())), Optional.of(node)));
+            }
+            if (joinDistributionType.canReplicate() && !mustRepartition(joinNode)) {
+                JoinNode node = joinNode.withDistributionType(REPLICATED);
+                possibleJoinNodes.add(new JoinEnumerationResult(Optional.of(lookup.getCumulativeCost(node, session, symbolAllocator.getTypes())), Optional.of(node)));
+                node = node.flipChildren();
+                possibleJoinNodes.add(new JoinEnumerationResult(Optional.of(lookup.getCumulativeCost(node, session, symbolAllocator.getTypes())), Optional.of(node)));
+            }
+
+            return resultOrdering.min(possibleJoinNodes);
+        }
+    }
+
+    @VisibleForTesting
+    static class JoinEnumerationResult
+    {
+        Optional<PlanNode> planNode;
```

Comment:
For cross joins you want to continue to the cross join phase.  for unknown cost you want to stop and return an empty result. You can't represent them both the same because you want to do something different with each.  In terms of representation, you don't want an infinite cost because that's a valid calculated cost.   I can do the opposite of your suggesetion though where for cross joins i'll return an empty result (since we haven't done anything yet).  and for uknown cost I'll leave it as is.  And I'll make the cost not optional.